src/backend/distributed/commands/table.c

```diff
@@ -819,23 +822,36 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand,
 		else if (alterTableType == AT_AttachPartition)
 		{
 			PartitionCmd *partitionCommand = (PartitionCmd *) command->def;
-
-			/*
-			 * We only support ALTER TABLE ATTACH PARTITION, if it is only subcommand of
-			 * ALTER TABLE. It was already checked in ErrorIfUnsupportedAlterTableStmt.
-			 */
-			Assert(list_length(commandList) <= 1);
-
 			rightRelationId = RangeVarGetRelid(partitionCommand->name, NoLock, false);
+			char rightRelationKind = get_rel_relkind(rightRelationId);
 
 			/*
-			 * Do not generate tasks if relation is distributed and the partition
-			 * is not distributed. Because, we'll manually convert the partition into
-			 * distributed table and co-locate with its parent.
+			 * We support ALTER INDEX ATTACH PARTITION and ALTER TABLE ATTACH PARTITION 
+			 * if it is only subcommand of ALTER TABLE command. Since the right relation
+			 * type is index for ALTER INDEX ATTACH PARTITION, we need to use the relation
+			 * id this index is created for.
+			 *
+			 * Both were already checked in ErrorIfUnsupportedAlterIndexStmt and
+			 * ErrorIfUnsupportedAlterTableStmt.
 			 */
-			if (!IsCitusTable(rightRelationId))
+			if (rightRelationKind == RELKIND_INDEX)
 			{
-				return NIL;
+				bool missingOk = false;
+				rightRelationId = IndexGetRelation(rightRelationId, missingOk);
```

Comment:
It is very implicit that we override the `rightRelationId` (a.k.a., `rightIndexOid`) to `rightRelationId`. Following such a change in the code is not trivial.

For example, just having a new variable for the output of `RangeVarGetRelid(partitionCommand->name, NoLock, false);` and using it would make things easier to follow. Can you improve this a bit?

Minor: We can probably have `Assert(IsCitusTable(rightRelationId))` because we already checked the left is distributed, and the right has to be distributed (with some comments in the code)

Also minor: `RelayEventExtendNamesForInterShardCommands` and `InterShardDDLTaskList` etc. all works fine because we already have the logic to handle `AT_ATTACH` for tables. Would be nice to update comments where relevant such as `/* if foreign key related, use specialized task list function ... */` where we could even mention that we pass the corresponding relationId instead of the indexRelId etc.

Reply:
Made it more explicit and updated comments on necessary parts.